spec/01-unit/01-db/01-schema/11-declarative_config/04-on-the-fly-migration_spec.lua

```diff
@@ -0,0 +1,119 @@
+require("spec.helpers") -- for kong.log
+local declarative = require "kong.db.declarative"
+local conf_loader = require "kong.conf_loader"
+
+local null = ngx.null
+
+local helpers = require "spec.helpers"
+local tablex = require "pl.tablex"
+local utils = require "kong.tools.utils"
+
+local function sort_by_key(t)
+  return function(a, b)
+    for _, k in ipairs({"name", "username", "host", "scope"}) do
+      local ka = t[a][k] ~= null and t[a][k]
+      local kb = t[b][k] ~= null and t[b][k]
+      if ka and kb then
+        return ka < kb
+      end
+    end
+  end
+end
+
+local function sortedpairs(t, fn)
+  local ks = tablex.keys(t)
+  table.sort(ks, fn and fn(t))
+  local i = 0
+  return function()
+    i = i + 1
+    return ks[i], t[ks[i]]
+  end
+end
+
+
+assert:set_parameter("TableFormatLevel", 10)
+
+
+local function idempotent(tbl, err)
+  assert.table(tbl, err)
+
+  for entity, items in sortedpairs(tbl) do
+    local new = {}
+    for _, item in sortedpairs(items, sort_by_key) do
+      table.insert(new, item)
+    end
+    tbl[entity] = new
+  end
+
+  local function recurse_fields(t)
+    helpers.deep_sort(t)
+    for k,v in sortedpairs(t) do
+      if k == "id" and utils.is_valid_uuid(v) then
+        t[k] = "UUID"
+      end
+      if k == "client_id" or k == "client_secret" or k == "access_token" then
+        t[k] = "RANDOM"
+      end
+      if type(v) == "table" then
+        recurse_fields(v)
+      end
+      if k == "created_at" or k == "updated_at" then
+        t[k] = 1234567890
+      end
+    end
+  end
+  recurse_fields(tbl)
+
+  table.sort(tbl)
+  return tbl
+end
+
+
+describe("declarative config: on the fly migration", function()
+  it("routes handling", function()
+    local dc = assert(declarative.new_config(conf_loader()))
+    local config = [[
+      _format_version: "1.1"
```

Comment:
make sure `2.1` is covered in tests, we can simply make it a loop and only change the version number